server/models/account/user.ts

```diff
@@ -354,6 +361,91 @@ export class UserModel extends Model<UserModel> {
     return UserModel.findOne(query)
   }
 
+  static loadByLdapDn (ldapDn: string) {
+    const query = {
+      where: {
+        ldapDn
+      }
+    }
+
+    return UserModel.findOne(query)
+  }
+
+  static findOrCreateLDAPUser (username: string) {
+    let userInfos
+
+    return Promise.resolve(UserModel.findLDAPUser(username))
+      .then((_userInfos) => {
+        userInfos = _userInfos
+        return UserModel.loadByLdapDn(userInfos['dn'])
+      })
+      .then((user) => {
+        if (user) {
+          return user
+        } else {
+          return UserModel.createLDAPUser(username, userInfos)
+        }
+      })
+      .catch(() => { return null })
+  }
+
+  static findLDAPUser (username: string) {
+    const client = ldap.createClient({
+      url: CONFIG.LDAP.URL,
+      tlsOptions: { rejectUnauthorized: !CONFIG.LDAP.INSECURE_TLS }
+    })
+    const filter = ldap.parseFilter(CONFIG.LDAP.USER_FILTER)
+    filter.forEach(function (element) {
+      if (element.value === '%username%') element.value = username
+    })
+    const opts = {
+      filter,
+      scope: 'sub',
+      attributes: [ CONFIG.LDAP.MAIL_ENTRY, 'dn' ]
+    }
+
+    return new Promise(function (resolve, reject) {
+      client.bind(CONFIG.LDAP.BIND_DN, CONFIG.LDAP.BIND_PASSWORD, function (err) {
+        if (err) reject()
+        let entries = []
+        client.search(CONFIG.LDAP.BASE, opts, function (err, search) {
+          if (err) reject()
+          search.on('searchEntry', function (entry) {
+            entries.push(entry.object)
+          })
+          search.on('end', function (result) {
+            if (entries.length === 1) {
+              resolve(entries[0])
+            } else {
+              reject()
+            }
+          })
+        })
+      })
+    })
+  }
+
+  static createLDAPUser (username: string, userInfos: {}) {
+    return Promise.resolve(userInfos)
+      .then((userInfos) => {
+        const userToCreate = new UserModel({
+          username: username,
+          password: 'SomeInvalidPassword',
```

Comment:
`null` no?

Reply:
password field cannot be null, and I didn’t want to remove this constraint as I don’t know the consequences in the rest of the code. If you think it’s safe, we can make it nullable

Reply:
Yep I think we can allow null on this column